include/tango/server/readers_writers_lock.h

```diff
@@ -34,83 +34,57 @@
 #include <thread>
 #include <atomic>
 
+namespace {
+    thread_local bool isWriter = false;
+};
 //TODO current implementation may lead to writer thread starvation, a lot of quick readers will always sneak while writer will wait till htey all gone
 class ReadersWritersLock {
     using thread = std::thread;
     using Lock = unique_lock<mutex>;
-public:
-    mutex mut;
-    condition_variable cond{};
-    int n;    // 0 means no-one active, > 0 means n readers, < 0 means writer
-    // (-n times).
-    thread::id writerId;
 
-    ReadersWritersLock(void) : n(0), writerId(nullptr) {}
+    mutex readersLock;
+    mutex writerLock;
+    condition_variable readers;
+    condition_variable writers;
+    atomic_int totalReaders;
+    atomic_int totalWriters;
+public:
+    ReadersWritersLock(void) : totalReaders(0), totalWriters(0) {}
 
     void readerIn(void) {
-        thread::id threadId{this_thread::get_id()};
-
+        if(totalWriters.load() > 0 && !::isWriter)
```

Comment:
Strangely enough but reading `::isWriter` in release mode always returns false, i.e. forces the thread which already has a WriterLock to wait forever. Looks like *::variable* does not refer to accessing variables in the unnamed namespaces, but rather smth totally different, see, for instance, [this post on SO](http://stackoverflow.com/questions/357404/why-are-unnamed-namespaces-used-and-what-are-their-benefits)